### DISCUSSION TITLE
Issue rendering singleband image with colormap via stac endpoint

### DISCUSSION THREAD

**Discussion by aliasmrchips** [developmentseed/titiler#235] [Q&A] [answered]
In particular, if I follow the rasterio [example](https://rasterio.readthedocs.io/en/latest/topics/color.html) and create a "colormap.tif". Loading the image directly using the cog endpoint works just fine. 
<img width="770" alt="Screen Shot 2021-02-15 at 6 32 27 PM" src="https://user-images.githubusercontent.com/963429/108007901-49587200-6fbc-11eb-8daa-f724a0a258bc.png">

However, if I create a STAC item and use the stac endpoints, it appears to ignore the internal colormap.
<img width="769" alt="Screen Shot 2021-02-15 at 6 33 01 PM" src="https://user-images.githubusercontent.com/963429/108007910-4cebf900-6fbc-11eb-8652-bcf05303c410.png">
I can't figure out what the difference is. Please help!

**Comment by vincentsarago** [comment, marked as answer]
Thanks @aliasmrchips for the question. 
You are right, the STACReader (used in the /stac) endpoint is loosing the information about the internal colormap of the item. This is by design and sadly there isn't an easy way of fixing this.

when you are using the `/cog` endpoint, you are using https://github.com/cogeotiff/rio-tiler/blob/master/rio_tiler/io/cogeo.py#L28 and https://github.com/developmentseed/titiler/blob/a95073d27ce0f8cb4699226173937dfe16c84b9e/titiler/endpoints/factory.py#L320-L322

☝️ in this, if the `COGReader` instance has a colormap variable, it will be used by default, and that's what you are getting.

For `/stac`, we are using a specific reader based on https://github.com/cogeotiff/rio-tiler/blob/master/rio_tiler/io/base.py#L334 which enable reading multiple assets in parallel (and merging the results), sadly with the STACReader we CANNOT keep the colormap for each asset 🙁. 

If the assets in your items do not need to be merged (which is what we do when multiple bands are stored as multiple assets), I would advise to switch to the `/cog/` endpoints.


Another solution would be to register a custom colormap (if the colormap in your data never change) and to use it to colorize your result 

1.  custom cmap: https://github.com/developmentseed/titiler/blob/master/titiler/custom/cmap.py#L10-L21
2. register a cmap to rio-tiler colormap https://github.com/developmentseed/titiler/blob/master/titiler/dependencies.py#L27
3. create a enum : https://github.com/developmentseed/titiler/blob/master/titiler/dependencies.py#L32-L34
4. create a dependency using your cmap https://github.com/developmentseed/titiler/blob/master/titiler/dependencies.py#L316-L343

Note: I'll update the docs to clearly explain how to create/register a custom colormap

**Reply by aliasmrchips** [reply]
Maybe you can clarify something for me... if I have checked out the repo (and deployed following the aws lambda directions) I won't actually see any changes if I redeploy because the lambda Dockerfile is actually installing titiler from pypi?

**Reply by vincentsarago** [reply]
absolutely, the AWS cdk code is an example, to deploy your own custom code using TiTiler you'll have to create some custom CDK code 
- https://github.com/developmentseed/titiler-pds
- https://github.com/developmentseed/titiler-mvt
- https://github.com/developmentseed/titiler-digitaltwin

**Comment by vincentsarago** [comment]
OR you could create your own `cog/stac` endpoint using a custom COG/STAC reader like 👇 

```python
import attr
from rasterio.io import DatasetReader
from rio_tiler.io.stac import fetch, _to_pystac_item
from rio_tiler.io import COGReader

import pystac

@attr.s
class CustomSTACReader(COGReader):
    """Custom COG Reader with GCPS support."""

    # This will keep the STAC item info within the instance
    item: pystac.Item = attr.ib(default=None, init=False)

    def __attrs_post_init__(self):
        """Define _kwargs, open dataset and get info."""
        # get STAC item URL and asset name
        asset = self.filepath.split(":")[-1]
        stac_url = self.filepath.replace(f":{asset}", "")

        # Fetch the STAC item
        self.item = pystac.Item.from_dict(fetch(stac_url), stac_url)

        # Get asset url from the STAC Item
        self.filepath = self.item.assets[asset].get_absolute_href()
        super().__attrs_post_init__()

with CustomSTACReader("https://canada-spot-ortho.s3.amazonaws.com/canada_spot_orthoimages/canada_spot5_orthoimages/S5_2007/S5_11055_6057_20070622/S5_11055_6057_20070622.json:pan") as cog:
    print(cog.bounds)

>>> (-111.87793996076493, 60.48627186654449, -109.94924666908423, 61.42036313093244)
```

then just create a custom tiler using the tiler factory 

```python
from titiler.endpoints.factory import TilerFactory

stac = TileFactory(reader=CustomSTACReader)

```

**Reply by aliasmrchips** [reply]
Thanks for the quick response and detailed explanation!

**Comment by vincentsarago** [comment]
FYI, in 0.2.0 we added custom colormap in the query parameter 
``` 
/stac/preview?assets=B1&colormap='{"0": "#000"....}'

**Reply by aliasmrchips** [reply]
Awesome! However, I can't seem to get it to work... do you have an exact example? I assume I am doing something wrong encoding the dictionary in the url.

**Reply by vincentsarago** [reply]
do you have an error message to share? 

you can also have a look at https://github.com/developmentseed/titiler/blob/287201a554523a1cb4258ff41ec52ca2bdc0ac13/tests/routes/test_cog.py#L225-L241

**Reply by aliasmrchips** [reply]
The error message is 
```
{
detail: "Expecting value: line 1 column 1 (char 0)"
}
```
or some variation.

Adapting the example above does work.

Thanks again!

**Reply by vincentsarago** [reply]
yeah, make sure the colormap is a json string and urlencoded

**Reply by aliasmrchips** [reply]
How would you pass in a gradient colormap, like "viridis" (if it was not a named colormap), for example?
